### DISCUSSION TITLE
Uncaught (in promise) Error: TodoList has not been deployed to detected network (network/artifact mismatch)

### DISCUSSION THREAD

**Discussion by Smit-05** [ConsenSys-archive/truffle#5226] [General Q&A]
## Issue

I am new to truffle and blockchain world and I have tried following a youtube video to make this project. But I am getting an error in deploying the contract.

***Uncaught (in promise) Error: TodoList has not been deployed to detected network (network/artifact mismatch)
    at truffle-contract.js:456:17
    at async Object.loadContract (app.js:60:20)
    at async Object.load (app.js:13:5)***

Forgive me for the rudeness as it is my first issue.
## TodoList.sol

```
pragma solidity ^0.5.0;

contract TodoList{
    uint public taskCount=0;

    struct Task{
        uint id;
        string taskName;
        bool completed;
    }

    mapping(uint => Task) public tasks;

    constructor () public {
        createTask("Start doig tasks now :)");
    }

    function createTask(string memory content) public {
        taskCount++;
        tasks[taskCount] = Task(taskCount,content,false);

    }
}
```

## app.js
```
var print = console.log

Web3 = require('web3')
App = {
  loading:false,
  contracts: {},

  load: async () => {
    await App.loadWeb3()
    await App.loadAccount()
    await App.loadContract()
    await App.render()
  },

  loadWeb3: async () => {
    
    if (typeof web3 !== 'undefined') {
      App.web3Provider = web3.currentProvider
      web3 = new Web3(web3.currentProvider)
    } else {
      window.alert("Please connect to Metamask.")
    }
    // Modern dapp browsers...
    if (window.ethereum) {
      window.web3 = new Web3(ethereum)
      try {
        // Request account access if needed
        await ethereum.enable()
        // Acccounts now exposed
        web3.eth.sendTransaction({/* ... */})
      } catch (error) {
        // User denied account access...
      }
    }
    // Legacy dapp browsers...
    else if (window.web3) {
      App.web3Provider = web3.currentProvider
      window.web3 = new Web3(web3.currentProvider)
      // Acccounts always exposed
      web3.eth.sendTransaction({/* ... */})
    }
    // Non-dapp browsers...
    else {
      console.log('Non-Ethereum browser detected. You should consider trying MetaMask!')
    }
  },

  loadAccount: async () => {
    App.account = window.web3.eth.accounts[0]
    print(App.account)
  },

  loadContract: async () => {
    const todoList = await $.getJSON('TodoList.json')
    App.contracts.TodoList = TruffleContract(todoList)
    App.contracts.TodoList.setProvider(App.web3Provider)
    print(todoList)
    App.todoList = await App.contracts.TodoList.deployed()
    print(App.todoList)
  },

  render: async () => {

    if(App.loading){
      return
    }
    App.setLoading(true)


    $("#account").html(App.account)

    App.setLoading(false)
  },


  setLoading: (boolean) => {
    App.loading = boolean
    const loader = $('#loader')
    const content = $('#content')
    if (boolean) {
      loader.show()
      content.hide()
    } else {
      loader.hide()
      content.show()
    }
  }

}

$(() => {
  $(window).load(() => {
    App.load()
  })
})
```
## 2_deploy_contracts.js
```
var TodoList = artifacts.require("./TodoList.sol");

module.exports = function(deployer) {
  deployer.deploy(TodoList);
};
```
## Environment

* Operating System: Windows 10
* Truffle version (`truffle version`):
Truffle v5.0.2 (core: 5.0.2)
Solidity v0.5.0 (solc-js)
* node version (`node --version`): v16.15.0
* npm version (`npm --version`): 8.12.1

**Comment by dongmingh** [comment]
@Smit-05 thanks for opening this. We move to discussion instead of issue so that our customer support can help you.

**Comment by apdan510** [comment]
@Smit-05 

Thanks for reaching out. It looks like the contract was not deployed on the network you are running. Could you please share which video tutorial you are following, your github repo of your project or your truffleconfig.js file? Any additional screenshots or logs are greatly appreciated. 

Kind regards,
Alex | **Truffle Support** | Consensys

**Reply by Smit-05** [reply]
Yeah sure!
I am following Dapp University tutorial for beginners.
Link: https://www.youtube.com/watch?v=rzvk2kdjr2I
MyRepo: https://github.com/Smit-05/BlockChain-ToDo-List

### truffle-config.js
```js
module.exports = {
  networks: {
    development: {
      host: "127.0.0.1",
      port: 7545,
      network_id: "*" // Match any network id
    },
  },
  solc: {
    optimizer: {
      enabled: true,
      runs: 200
    }
  }
}
```

Thankyou once again for helping out.

**Reply by apdan510** [reply]
Nice, I remember this exact tutorial from DappU. This tutorial is a bit more dated now, but I am happy to assist here. I just tested compiling, migrating, and deploying your contracts and everything looks good on my end. When are you getting these errors? What exact commands are you entering into your terminal to return that error. 

Kind regards,
Alex | Truffle Support | Consensys

**Reply by Smit-05** [reply]
It is working completely fine in terminal. But when I start the server and run it on browser, then it would show me this error in console. 
- I used `npm run dev` command to start server.
- As I was getting a Web3 error i.e. ```Uncaught TypeError: Web3 is not a constructor```.
- I used this script to solve that in index.html ```<script src="https://cdn.jsdelivr.net/gh/ethereum/web3.js@0.20.6/dist/web3.min.js"></script>```. I saw this solution somewhere on stackoverflow.
- After that I got that error of contract not deployed
![image](https://user-images.githubusercontent.com/85776968/175759361-c52f7fd9-58b6-42f7-afa9-d5a018ce3ac2.png)

Thankyou once again :)

**Reply by sampriti026** [reply]
were you able to debug it, i am facing the same error in another dappu tutorial.

**Reply by sampriti026** [reply]
So I was able to fix mine by replacing the filename with the contract class name in the 2_deploy_contracts.js file
like this, 
`var TodoList = artifacts.require("./TodoList.sol");` to `var TodoList = artifacts.require("TodoList");`

**Reply by apdan510** [reply]
@sampriti026 were there multiple contracts with the same name in the directory? Or was this file just renamed?

Thank you!

**Reply by Baharkaviani** [reply]
> It is working completely fine in terminal. But when I start the server and run it on browser, then it would show me this error in console.
> 
> * I used `npm run dev` command to start server.
> * As I was getting a Web3 error i.e. `Uncaught TypeError: Web3 is not a constructor`.
> * I used this script to solve that in index.html `<script src="https://cdn.jsdelivr.net/gh/ethereum/web3.js@0.20.6/dist/web3.min.js"></script>`. I saw this solution somewhere on stackoverflow.
> * After that I got that error of contract not deployed
>   ![image](https://user-images.githubusercontent.com/85776968/175759361-c52f7fd9-58b6-42f7-afa9-d5a018ce3ac2.png)
> 
> Thankyou once again :)

Hi @Smit-05,
I applied exactly the same changes to the project and ran into the same problem.
I found that the Metamask account was no longer correct for me due to restarting Ganache. After importing the correct account, the problem was solved.

Hope this helps :)

**Reply by anchengyang** [reply]
Hi! I would also like to ask for help regarding this issue as I faced the same problem of Web3 is not a constructor and I applied the same changes, after which I got the same error that the contract was not deployed to the detected network. The correct account has been imported into Metamask from Ganache so I am not sure what is wrong.

Let me know if you need any screenshots that may help

**Comment by Kyojur0** [comment]
same issue have you found out your problem ? if so plz lemme know

**Reply by apdan510** [reply]
@Kyojur0 Can you share a screenshot or share your repo? I want to see how the directory is set up. 

Thanks!

**Reply by Kyojur0** [reply]
sure , the test directory is empty at the moment 
![image](https://user-images.githubusercontent.com/89397672/179953561-0f34008e-4640-4c8a-b82d-74e9ccdd153c.png)

**Reply by Kyojur0** [reply]
https://github.com/Kyojur0/Eth-ToDoList.git

heres the link if you want to see the code for yourself

**Reply by apdan510** [reply]
@Kyojur0 can you layout the steps you took on deployment of the todo contract? How did you deploy it? Was in it the terminal, was ganache ui running? was the deployment or migration successful?

**Reply by Kyojur0** [reply]
sure well first i did 
`truffle compile`
it works fine generates output in build folder 

then `truffle migrate --network local_network`
```
PS C:\Users\Rias Gremory\Desktop\eth-todo-list> truffle migrate --network local_network
⚠️  Important ⚠️
If you're using an HDWalletProvider, it must be Web3 1.0 enabled or your migration will hang.


Starting migrations...
======================
> Network name:    'local_network'
> Network id:      5777
> Block gas limit: 6721975


1_initial_migration.js
======================

   Deploying 'Migrations'
   ----------------------
   > transaction hash:    0xbdf78a72773cea08c7170db8454c71bdb0901e4bcfbb8d2147c5cbd8678f6d4e
   > Blocks: 0            Seconds: 0
   > contract address:    0x3E02244dbC6f4B7518953bACb8bBC53Ed17f8FC2
   > account:             0x64aa943e00acbF07cd41a52325759E6b51567201
   > balance:             99.93691746
   > gas used:            179719
   > gas price:           20 gwei
   > value sent:          0 ETH
   > total cost:          0.00359438 ETH

   > Saving artifacts
   -------------------------------------
   > Total cost:          0.00359438 ETH


2_Deploy_Contract.js
====================

   Replacing 'ToDoList'
   --------------------
   > transaction hash:    0xd53423a1af0752f640980b4049230ec8145dc15476f1b23d80e03e1aa845234f
   > Blocks: 0            Seconds: 0
   > contract address:    0xb79cB67fCe87E7FDc675cf666BA62C20200CA09b
   > account:             0x64aa943e00acbF07cd41a52325759E6b51567201
   > balance:             99.9299596
   > gas used:            347893
   > gas price:           20 gwei
   > value sent:          0 ETH
   > total cost:          0.00695786 ETH

   > Saving artifacts
   -------------------------------------
   > Total cost:          0.00695786 ETH


Summary
=======
> Total deployments:   2
> Final cost:          0.01055224 ETH
```
which also works fine some of the ethers do get deducted from my Ganache Account (GUI)

**Reply by Kyojur0** [reply]
i also tried somethings like `truffle deploy --network local_network`
which gives same results as migrate one 

![image](https://user-images.githubusercontent.com/89397672/180508229-b6078546-21e9-40aa-ad13-3b7a68f2eda9.png)
![image](https://user-images.githubusercontent.com/89397672/180508487-6958e276-3f2f-41bd-8319-dc744c9e3f5b.png)
 
this is latest ganache transitionary I did right now

**Reply by Kyojur0** [reply]
lmk if you need something else 
and thank you :) <3

**Reply by LoveKosmos** [reply]
Hi, @Kyojur0 could you try applying changes that were suggested in this message [https://github.com/trufflesuite/truffle/discussions/5226#discussioncomment-3046625](url)

Let me know if that worked for you!

**Comment by prakhar-da-gama** [comment]
press **ctrl+c** in MinGW terminal
use command **truffle migrate**
then **truffle console**
then **todoList = await TodoList.deployed()**

it should work now
